src/WebJobs.Script.WebHost/HostNameProvider.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    /// <summary>
+    /// Provides the current HostName for the Function App.
+    /// <remarks>
+    /// The environment value for WEBSITE_HOSTNAME is unreliable and shouldn't be used directly. AppService site swaps change
+    /// the site’s hostname under the covers, and the worker process is NOT recycled (for performance reasons). That means the
+    /// site will continue to run with the same hostname environment variable, leading to an incorrect host name.
+    ///
+    /// WAS_DEFAULT_HOSTNAME is a header injected by front end on every request which provides the correct hostname. We check
+    /// this header on all http requests, and updated the cached hostname value as needed.
+    /// </remarks>
+    /// </summary>
+    public class HostNameProvider
+    {
+        private readonly IEnvironment _environment;
+        private readonly ILogger _logger;
+        private string _hostName;
+
+        public HostNameProvider(IEnvironment environment, ILoggerFactory loggerFactory)
+        {
+            _environment = environment ?? throw new ArgumentException(nameof(environment));
+            if (loggerFactory == null)
+            {
+                throw new ArgumentException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger(ScriptConstants.LogCategoryHostGeneral);
```

Comment:
nit: rather than `LoggerFactory`, take an `ILogger<HostNameProvider>` in the constructor. It'll get injected and gives us better categories for filtering.